src/elements/a-gltf-entity.js

```diff
@@ -0,0 +1,190 @@
+
+  for (var i = 0; i < targetEls.length; i++) {
+    const targetEl = targetEls[i];
+
+    // Merge root element attributes with the target element
+    for (var i = 0; i < templateRootAttrs.length; i++) {
+      targetEl.setAttribute(templateRootAttrs[i].name, templateRootAttrs[i].value);
+    }
+
+    // Append all child elements
+    for (var i = 0; i < templateRoot.children.length; i++) {
+      targetEl.appendChild(document.importNode(templateRoot.children[i], true));
+    }
+  }
+}
+
+AFRAME.registerElement("a-gltf-entity", {
+  prototype: Object.create(AFRAME.AEntity.prototype, {
+    load: {
+      value() {
+        if (this.hasLoaded || !this.parentEl) {
+          return;
+        }
+
+        // Get the src url.
+        let src = this.getAttribute("src");
+
+        // If the src attribute is a selector, get the url from the asset item.
+        if (src.charAt(0) === "#") {
+          const assetEl = document.getElementById(src.substring(1));
+          src = assetEl.getAttribute("src");
+        }
+
+        // Load the gltf model from the cache if it exists.
+        const gltf = GLTFCache[src];
+
+        if (gltf) {
+          // Use a cloned copy of the cached model.
+          const clonedGltf = cloneGltf(gltf);
+          this.onLoad(clonedGltf);
+          return;
+        }
+
+        const finalizeLoad = () => {
+          AFRAME.ANode.prototype.load.call(this, () => {
+            // Check if entity was detached while it was waiting to load.
+            if (!this.parentEl) {
+              return;
+            }
+
+            this.updateComponents();
+            if (this.isScene || this.parentEl.isPlaying) {
+              this.play();
+            }
+          });
+        };
+
+        const inflate = (model, callback) => {
+          inflateEntities("", this, model);
+          this.querySelectorAll(":scope > template").forEach(attachTemplate);
+          setTimeout(callback, 0);
+        };
+
+        const onLoad = gltfModel => {
+          if (!GLTFCache[this.data]) {
+            // Store a cloned copy of the gltf model.
+            GLTFCache[this.data] = cloneGltf(gltfModel);
+          }
+
+          this.model = gltfModel.scene || gltfModel.scenes[0];
+          this.model.animations = gltfModel.animations;
+
+          this.setObject3D("mesh", this.model);
+          this.emit("model-loaded", { format: "gltf", model: this.model });
+
+          if (this.getAttribute("inflate")) {
+            inflate(this.model, finalizeLoad);
+          } else {
+            finalizeLoad();
+          }
+        };
+
+        const onError = error => {
+          const message = error && error.message ? error.message : "Failed to load glTF model";
+          console.warn(message);
+          this.emit("model-error", { format: "gltf", src: this.data });
+        };
+
+        // Otherwise load the new gltf model.
+        new THREE.GLTFLoader().load(src, onLoad, undefined /* onProgress */, onError);
```

Comment:
Maybe just personal preference but id inline `onLoad` and `onError` here as I find the current way this reads a bit confusing to follow the flow of, also I confused `this.onLoad` with `onLoad` at first.